h5py/h5t.pyx

```diff
@@ -1421,27 +1421,61 @@ cdef TypeCompoundID _c_complex(dtype dt):
 
     return TypeCompoundID(tid)
 
-cdef TypeCompoundID _c_compound(dtype dt, int logical):
+cdef TypeCompoundID _c_compound(dtype dt, int logical, int aligned):
     # Compound datatypes
 
     cdef hid_t tid
     cdef TypeID type_tmp
     cdef dtype dt_tmp
     cdef size_t offset
+    cdef size_t offset_step = 0
 
-    cdef dict fields = dt.fields
     cdef tuple names = dt.names
-
-    # Initial size MUST be 1 to avoid segfaults (issue 166)
-    tid = H5Tcreate(H5T_COMPOUND, 1)  
-
-    for name in names:
-        ename = name.encode('utf8') if isinstance(name, unicode) else name
-        dt_tmp = dt.fields[name][0]
-        offset = dt.fields[name][1]
-        type_tmp = py_create(dt_tmp, logical=logical)
-        H5Tset_size(tid, offset+type_tmp.get_size())
-        H5Tinsert(tid, ename, offset, type_tmp.id)
+    cdef dict fields = {}
+    cdef list offsets
+
+    # The challenge with correctly converting a numpy/h5py dtype to a HDF5 type
+    # which is composed of subtypes has three aspects we must consider
+    # 1. numpy/h5py dtypes do not always have the same size and HDF5, even when
+    #   equivalent (can result in overlapping elements if not careful)
+    # 2. For correct round-tripping of aligned dtypes, we need to consider how
+    #   much padding we need
+    # 3. There is no requirement that the offsets be monotonically increasing
+    #
+    # The code below tries to cover these aspects
+
+    for name, field in dt.fields.items():
+        dt_tmp = field[0]
+        offset = field[1]
+        fields[offset] = {
+            "name": name.encode('utf8') if isinstance(name, unicode) else name,
```

Comment:
The numpy C-API is unclear about the types here (or at least the documentation is), so I'm not sure we want to change this.